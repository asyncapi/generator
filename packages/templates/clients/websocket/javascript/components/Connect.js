import { Text } from '@asyncapi/generator-react-sdk';

export function Connect({ title }) {
  return (
    <Text newLines={2} indent={2}>
      {
        `// Method to establish a WebSocket connection
connect() {
  return new Promise((resolve, reject) => {
    this.websocket = new WebSocket(this.url);

    // On successful connection
    this.websocket.onopen = () => {
      console.log('Connected to ${title} server');
      resolve();
    };

    // On receiving a message
    this.websocket.onmessage = (event) => {
      if (this.messageHandlers.length > 0) {
        // Call custom message handlers
        this.messageHandlers.forEach(handler => {
          if (typeof handler === 'function') {
            this.handleMessage(event.data, handler);
          }
        });
      } else {
        // Default message logging
        console.log('Message received:', event.data);
      }
    };

    // On error first call custom error handlers, then default error behavior
    this.websocket.onerror = (error) => {
      if (this.errorHandlers.length > 0) {
        // Call custom error handlers
        this.errorHandlers.forEach(handler => handler(error));
      } else {
        // Default error behavior
        console.error('WebSocket Error:', error);
      }
      reject(error);
    };

    // On connection close
    this.websocket.onclose = () => {
      console.log('Disconnected from ${title} server');
    };
  });
}`
      }
    </Text>
  );
}
