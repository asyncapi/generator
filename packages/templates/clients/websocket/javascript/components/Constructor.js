import { Text } from '@asyncapi/generator-react-sdk';

export function Constructor({ serverUrl }) {
  return (
    <Text indent={2}>
      {
        `/*
  * Constructor to initialize the WebSocket client
  * @param {string} url - The WebSocket server URL. Use it if the server URL is different from the default one taken from the AsyncAPI document.
*/
constructor(url) {
  this.url = url || '${serverUrl}';
  this.websocket = null;
  this.messageHandlers = [];
  this.errorHandlers = [];
}
`
      }
    </Text>
  );
}
