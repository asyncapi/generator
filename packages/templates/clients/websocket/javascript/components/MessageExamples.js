import { getMessageExamples, getOperationMessages } from '@asyncapi/generator-helpers';
import { Text } from '@asyncapi/generator-react-sdk';

export default function MessageExamples({operation}) {
  const operationId = operation.id();
  const messageExamples = [];
  const messages = getOperationMessages(operation) || [];

  messages.forEach((message) => {
    const examples = getMessageExamples(message) || [];
    examples.forEach((example) => {
      const payload = example.payload();
      messageExamples.push(`\n\n**Example:**\n\`\`\`javascript\nclient.${operationId}(${JSON.stringify(payload, null, 2)});\n\`\`\``);
    });
  });
  return (
    <Text>
      {messageExamples.map(example => (
        <Text>
          {example}
        </Text>
      ))}
    </Text>
  );
}