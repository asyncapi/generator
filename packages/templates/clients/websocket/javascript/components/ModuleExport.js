import { Text } from '@asyncapi/generator-react-sdk';

export function ModuleExport({ clientName }) {
  return (
    <Text>
      {
        `module.exports = ${clientName};`
      }
    </Text>
  );
}
