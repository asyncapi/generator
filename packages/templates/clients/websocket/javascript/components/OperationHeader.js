import { Text } from '@asyncapi/generator-react-sdk';

export default function OperationHeader({operation}) {
  const operationId = operation.id();
  return (
    <Text>
      {`#### \`${operationId}(payload)\`
${operation.hasSummary() ?  operation.summary() : ''}
${operation.hasDescription() ? `\n${operation.description()}` : ''}`}
    </Text>
  );
}