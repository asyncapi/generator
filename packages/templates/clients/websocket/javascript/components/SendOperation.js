import { Text } from '@asyncapi/generator-react-sdk';

export function SendOperation({ sendOperations, clientName }) {
  if (!sendOperations || sendOperations.length === 0) {
    return null;
  }

  return (
    <>
      {sendOperations.map((operation) => (
        <Text newLines={2} indent={2}>
          {`
/**
 * Sends a ${operation.id()} message over the WebSocket connection.
 * 
 * @param {Object} message - The message payload to send. Should match the schema defined in the AsyncAPI document.
 * @param {WebSocket} [socket] - The WebSocket connection to use. If not provided, the client's own connection will be used.
 * @throws {TypeError} If message cannot be stringified to JSON
 * @throws {Error} If WebSocket connection is not in OPEN state
 */
static ${operation.id()}(message, socket) {
  try {
    socket.send(JSON.stringify(message));
  } catch (error) {
    console.error('Error sending ${operation.id()} message:', error);
  }
}
/**
 * Instance method version of ${operation.id()} that uses the client's own WebSocket connection.
 * @param {Object} message - The message payload to send
 * @throws {Error} If WebSocket connection is not established
 */
${operation.id()}(message){
  if(!this.websocket){
    throw new Error('WebSocket connection not established. Call connect() first.');
  }
  ${clientName}.${operation.id()}(message, this.websocket);
}
  `}
        </Text>
      ))}
    </>
  );
}
