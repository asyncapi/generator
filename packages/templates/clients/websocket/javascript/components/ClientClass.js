import { Text } from '@asyncapi/generator-react-sdk';
import { Constructor } from './Constructor';
import { Connect } from './Connect';
import { RegisterMessageHandler } from './RegisterMessageHandler';
import { RegisterErrorHandler } from './RegisterErrorHandler';
import { HandleMessage } from './HandleMessage';
import { SendEchoMessage } from './SendEchoMessage';
import { CloseConnection } from './CloseConnection';
import { ModuleExport } from './ModuleExport';

export function ClientClass({ clientName, serverUrl, title }) {
  return (
    <Text>
      <Text newLines={2}>
        {`class ${clientName} {`}
      </Text>
      <Constructor serverUrl={serverUrl} />
      <Connect title={title} />
      <RegisterMessageHandler />
      <RegisterErrorHandler />
      <HandleMessage />
      <SendEchoMessage />
      <CloseConnection />
      <Text>
        {'}'}
      </Text>
      <ModuleExport clientName={clientName} />
    </Text>
  );
}
