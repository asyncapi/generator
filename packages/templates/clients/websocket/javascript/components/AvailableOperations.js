import { Text } from '@asyncapi/generator-react-sdk';
import OperationHeader from './OperationHeader';
import MessageExamples from './MessageExamples';

export function AvailableOperations({ operations }) {
  return (
    <>
      <Text newLines={2}>### Available Operations</Text>
      {operations.map((operation) => (
        <Text newLines={2}>
          <OperationHeader operation={operation} />
          <MessageExamples operation={operation} />
        </Text>
      ))}
    </>
  );
}
