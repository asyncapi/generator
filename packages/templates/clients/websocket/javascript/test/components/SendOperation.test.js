import path from 'path';
import { render } from '@asyncapi/generator-react-sdk';
import { Parser, fromFile } from '@asyncapi/parser';
import { SendOperation } from '../../components/SendOperation';

const parser = new Parser();
const asyncapi_websocket_query = path.resolve(__dirname, '../../../../../../helpers/test/__fixtures__/asyncapi-websocket-query.yml');

describe('Testing of SendOperation function', () => {
  let parsedAsyncAPIDocument;

  beforeAll(async () => {
    const parseResult = await fromFile(parser, asyncapi_websocket_query).parse();
    parsedAsyncAPIDocument = parseResult.document;
  });

  test('render websockets with send operations and client name', () => {
    const result = render(
      <SendOperation 
        clientName='GeminiMarketDataWebsocketAPI'
        sendOperations={parsedAsyncAPIDocument.operations().filterBySend()} 
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render websockets without send operations', () => {
    const result = render(
      <SendOperation 
        clientName='GeminiMarketDataWebsocketAPI'
        sendOperations={null} 
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });
});
