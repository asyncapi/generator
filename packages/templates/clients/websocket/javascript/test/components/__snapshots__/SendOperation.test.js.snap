// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing of SendOperation function render websockets with send operations and client name 1`] = `
"/**
   * Sends a noMessage message over the WebSocket connection.
   * 
   * @param {Object} message - The message payload to send. Should match the schema defined in the AsyncAPI document.
   * @param {WebSocket} [socket] - The WebSocket connection to use. If not provided, the client's own connection will be used.
   * @throws {TypeError} If message cannot be stringified to JSON
   * @throws {Error} If WebSocket connection is not in OPEN state
   */
  static noMessage(message, socket) {
    try {
      socket.send(JSON.stringify(message));
    } catch (error) {
      console.error('Error sending noMessage message:', error);
    }
  }
  /**
   * Instance method version of noMessage that uses the client's own WebSocket connection.
   * @param {Object} message - The message payload to send
   * @throws {Error} If WebSocket connection is not established
   */
  noMessage(message){
    if(!this.websocket){
      throw new Error('WebSocket connection not established. Call connect() first.');
    }
    GeminiMarketDataWebsocketAPI.noMessage(message, this.websocket);
  }
  

  /**
   * Sends a noSummaryOperations message over the WebSocket connection.
   * 
   * @param {Object} message - The message payload to send. Should match the schema defined in the AsyncAPI document.
   * @param {WebSocket} [socket] - The WebSocket connection to use. If not provided, the client's own connection will be used.
   * @throws {TypeError} If message cannot be stringified to JSON
   * @throws {Error} If WebSocket connection is not in OPEN state
   */
  static noSummaryOperations(message, socket) {
    try {
      socket.send(JSON.stringify(message));
    } catch (error) {
      console.error('Error sending noSummaryOperations message:', error);
    }
  }
  /**
   * Instance method version of noSummaryOperations that uses the client's own WebSocket connection.
   * @param {Object} message - The message payload to send
   * @throws {Error} If WebSocket connection is not established
   */
  noSummaryOperations(message){
    if(!this.websocket){
      throw new Error('WebSocket connection not established. Call connect() first.');
    }
    GeminiMarketDataWebsocketAPI.noSummaryOperations(message, this.websocket);
  }
  

  /**
   * Sends a noSummaryNoDescriptionOperations message over the WebSocket connection.
   * 
   * @param {Object} message - The message payload to send. Should match the schema defined in the AsyncAPI document.
   * @param {WebSocket} [socket] - The WebSocket connection to use. If not provided, the client's own connection will be used.
   * @throws {TypeError} If message cannot be stringified to JSON
   * @throws {Error} If WebSocket connection is not in OPEN state
   */
  static noSummaryNoDescriptionOperations(message, socket) {
    try {
      socket.send(JSON.stringify(message));
    } catch (error) {
      console.error('Error sending noSummaryNoDescriptionOperations message:', error);
    }
  }
  /**
   * Instance method version of noSummaryNoDescriptionOperations that uses the client's own WebSocket connection.
   * @param {Object} message - The message payload to send
   * @throws {Error} If WebSocket connection is not established
   */
  noSummaryNoDescriptionOperations(message){
    if(!this.websocket){
      throw new Error('WebSocket connection not established. Call connect() first.');
    }
    GeminiMarketDataWebsocketAPI.noSummaryNoDescriptionOperations(message, this.websocket);
  }
  

  /**
   * Sends a mixedMessageExamples message over the WebSocket connection.
   * 
   * @param {Object} message - The message payload to send. Should match the schema defined in the AsyncAPI document.
   * @param {WebSocket} [socket] - The WebSocket connection to use. If not provided, the client's own connection will be used.
   * @throws {TypeError} If message cannot be stringified to JSON
   * @throws {Error} If WebSocket connection is not in OPEN state
   */
  static mixedMessageExamples(message, socket) {
    try {
      socket.send(JSON.stringify(message));
    } catch (error) {
      console.error('Error sending mixedMessageExamples message:', error);
    }
  }
  /**
   * Instance method version of mixedMessageExamples that uses the client's own WebSocket connection.
   * @param {Object} message - The message payload to send
   * @throws {Error} If WebSocket connection is not established
   */
  mixedMessageExamples(message){
    if(!this.websocket){
      throw new Error('WebSocket connection not established. Call connect() first.');
    }
    GeminiMarketDataWebsocketAPI.mixedMessageExamples(message, this.websocket);
  }"
`;

exports[`Testing of SendOperation function render websockets without send operations 1`] = `""`;
